### PR TITLE
v2: Improve OAuth2Client error handling and add expand b2clogin allow list

### PR DESCRIPTION
**Pull request by henrik-me**
Better error handling for OAuth2Client and add support for B2Clogin.cn

### CODE REVIEW

File: src/Microsoft.Identity.Client/Instance/B2CAuthority.cs [hunk 1]
```diff
@@ -37,7 +37,7 @@ internal class B2CAuthority : AadAuthority
         public const string Prefix = "tfp"; // The http path of B2C authority looks like "/tfp/<your_tenant_name>/..."
         public const string B2CCanonicalAuthorityTemplate = "https://{0}/{1}/{2}/{3}/";
         public const string OpenIdConfigurationEndpoint = "v2.0/.well-known/openid-configuration";
-        public const string B2CTrustedHost = "b2clogin.com";
+        private string[] B2CTrustedHosts = { "b2clogin.com", "b2clogin.cn" };
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/Instance/B2CAuthority.cs]: what about BlackForest and US gov? Isn't B2C yet deployed there?

File: src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 2]
```diff
@@ -168,40 +178,57 @@ public static void CreateErrorResponse(HttpResponse response, RequestContext req
...
+                serviceEx = MsalExceptionFactory.GetServiceException(
+                    CoreErrorCodes.NonParsableOAuthError,
+                    CoreErrorMessages.NonParsableOAuthError,
+                    response);
             }
-            catch (SerializationException ex)
+            catch (Exception ex)
             {
-                serviceEx = MsalExceptionFactory.GetClientException(CoreErrorCodes.UnknownError, response.Body, ex);
+                serviceEx = MsalExceptionFactory.GetServiceException(CoreErrorCodes.UnknownError, response.Body, response, ex);
```
Comment (jennyf19) [inline src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: should be this:
```CSharp serviceEx = MsalExceptionFactory.GetServiceException(
                    CoreErrorCodes.UnknownError, 
                    response.Body, 
                    ex,
                    ExceptionDetail.FromHttpResponse(response));

File: src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 3]
```diff
@@ -168,40 +178,57 @@ public static void CreateErrorResponse(HttpResponse response, RequestContext req
...
+            }
+            catch (SerializationException) // in the rare case we get an error response we cannot deserialize
+            {
+                serviceEx = MsalExceptionFactory.GetServiceException(
+                    CoreErrorCodes.NonParsableOAuthError,
+                    CoreErrorMessages.NonParsableOAuthError,
+                    response);
             }
-            catch (SerializationException ex)
+            catch (Exception ex)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: The above MsalUiRequiredException is getting caught here, which is why all the tests are having issues for UiRequired stuff showing invalid_grant instead. Maybe that's the same w/dev3x? Currently we're just catching SerializationExceptions, but not all exceptions.

File: src/Microsoft.Identity.Client/Exceptions/CoreErrorCodes.cs [hunk 4]
```diff
@@ -75,6 +75,8 @@ internal class CoreErrorCodes
 
         public const string GetUserNameFailed = "get_user_name_failed";
         public const string NonParsableOAuthError = "non_parsable_oauth_error";
+        public const string HttpStatusNotFound = "not_found";
+        public const string HttpStatusCodeNotOk = "http_status_not_200";
```
Comment (jennyf19) [inline src/Microsoft.Identity.Client/Exceptions/CoreErrorCodes.cs]: These have to get added to MsalErrorCodes

File: src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs (hunk 3 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: will investigate.

File: src/Microsoft.Identity.Client/Exceptions/CoreErrorCodes.cs (hunk 4 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/Exceptions/CoreErrorCodes.cs]: will investigate.

File: src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs (hunk 2 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs]: possibly :)

File: src/Microsoft.Identity.Client/Instance/B2CAuthority.cs (hunk 1 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/Instance/B2CAuthority.cs]: Will follow up with B2C to get clarification on those.

File: tests/Microsoft.Identity.Test.Common/Core/Helpers/CoreTestConstants.cs [hunk 5]
```diff
@@ -48,6 +48,9 @@ public static class CoreTestConstants
         public const string B2CAuthority = "https://login.microsoftonline.in/tfp/tenant/policy/";
         public const string B2CLoginAuthority = "https://sometenantid.b2clogin.com/tfp/sometenantid/policy/";
         public const string B2CRandomHost = "https://sometenantid.randomhost.com/tfp/sometenantid/policy/";
+        public const string B2CLoginAuthorityUsGov = "https://sometenantid.b2clogin.us/tfp/sometenantid/policy/";
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Common/Core/Helpers/CoreTestConstants.cs]: we can probably clean these up at some point. move the tenantid, tfp, and policy to constants and then form the authority.

**Comment by henrik-me** [comment]
UI Automation. Android tests are not running. [All other tests are green}(https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=291610).

**Pull request merged**